client/service_request.go

```diff
@@ -155,3 +163,16 @@ func checkResponse(resp *http.Response) error {
 	}
 	return &errors.Error{StatusCode: resp.StatusCode, Message: string(bodyBytes)}
 }
+
+// storeRawBody reads a response body from b, stores it into bh, and returns a
+// new io.Reader that can be used for reading the response body a second time.
+func storeRawBody(bh rawBodyHolder, b io.Reader) (io.Reader, error) {
+	body, err := io.ReadAll(b)
+	if err != nil {
+		return nil, fmt.Errorf("reading response body: %w", err)
+	}
+	if _, err = bh.storeBody(bytes.NewReader(body)); err != nil {
+		return nil, fmt.Errorf("storing response body: %w", err)
+	}
+	return bytes.NewReader(body), nil
+}
```

Comment:
Can this be migrated and be reused everywhere? I guess yes?

Reply:
It doesn't even need to, because it applies to the generic `ServiceResponse` type. The only place we need to call this function is inside `DoRequest`.